### ISSUE TITLE
imshow in 2.1: color resolution depends on outliers? 

### ISSUE THREAD

**Issue by trettberg** [matplotlib/matplotlib#10567]
Since 2.1 the effective color resolution of `imshow(x, vmin=-1, vmax=1)` seems to depend on large outliers in `x`.

Is this intended behavior?

Thanks for clarification!

```python
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt

x = np.linspace(-1, 1, 500 )
x = np.ones([20, 1]) * x[np.newaxis, :]

x1 = x.copy()
x1[0, 0] = 1e16

x2 = x.copy()
x2[0, 0] = 1e17

_, axes = plt.subplots(nrows=3)
axes[0].imshow(x, vmin=-1, vmax=1)
axes[1].imshow(x1, vmin=-1, vmax=1)
axes[2].imshow(x2, vmin=-1, vmax=1)

axes[0].set_title(mpl.__version__)
plt.show()
```

**Actual outcome**
![figure_1-1](https://user-images.githubusercontent.com/16763319/36547074-43424b90-17ed-11e8-8041-5dea944f1097.png)

**Expected outcome**
![figure_1](https://user-images.githubusercontent.com/16763319/36547088-4b786358-17ed-11e8-8b33-47ccf90ad3e2.png)

**Matplotlib version**
  * Operating system: Linux
  * Matplotlib version: 2.1.0 (expected behavior in 2.0.2)
  * Matplotlib backend: Qt5Agg
  * Python version: 3.6.4
  * Other libraries: numpy 1.13.3
  * via conda

**Comment by dstansby** [comment]
Hmm, I'm not sure about whether this is intended or not, it bisects to https://github.com/matplotlib/matplotlib/commit/12c27f3892931f0f3113e3041bed38d195b1d40d so maybe @tacaswell will know more?

**Comment by WeatherGod** [comment]
Comment in that new code:
```
# This may introduce numeric instabilities in very broadly
# scaled data
```
Looks like we might have found that situation. The data is getting scaled
for the purposes of the AGG renderer (somewhat independent of the norm that
was specified). Unfortunately, due to some fun effects with floating point
numbers, that normalization gets very digitized when the range is very
large. I would say that this is unintentional, but I am not sure how we can
fix it, though.

**Comment by jklymak** [comment]
We messed around w/ all of this for 2.1.2.  #10133 W/o looking carefully, does that fix this?

**Comment by jklymak** [comment]
Pretty sure it does - it was a round off error that caused this, but its now been fixed for large floats.

**Comment by dstansby** [comment]
It didn't seem fixed when I ran the above test on the master branch.

**Comment by jklymak** [comment]
agreed.    The image is stored as float64.  You lose resolution at 2^53=9e15.  I’m not sure how this worked in 2.1, but that’s the problem here.

**Comment by WeatherGod** [comment]
the rescaling is relatively recent, iirc, as part of the work to better
handle over/under/invalid.

**Comment by jklymak** [comment]
I tried to set the dtype to float 128 fro big numbers, but I was told there is no such thing...

**Comment by tacaswell** [comment]
Image interpolation is the bug that just keeps giving!

The chain of bugs and fixes here goes:

 - while doing 2.0 we noticed that we were interpolating after doing norm + rgb mapping -> leads to colors not in the color map in the final image so we moved the interpolation to be done on the normed, but not color mapped data. 
 - this introduced https://github.com/matplotlib/matplotlib/issues/8012,  so we better masked around 'invalid' pixels when color mapping https://github.com/matplotlib/matplotlib/pull/8024
 - as a consequence of this over/under/invlaid pixels effectively poisoned everything in their  in their kernel window (https://github.com/matplotlib/matplotlib/issues/8631) so we moved interpolation to the raw data (https://github.com/matplotlib/matplotlib/pull/8966)
 - we now have numerics issues for data with large dynamic range (this bug + https://github.com/matplotlib/matplotlib/issues/10072 which was fixed by https://github.com/matplotlib/matplotlib/pull/10133)

**Comment by jklymak** [comment]
It’s the same bug, it’s just that #10133 only fixed up to 10^16 or so.

**Comment by jklymak** [comment]
Is there some reason to not apply the norm (but not RGB mapping) first?  If I do that, it seems to work for this case.  It fails a bunch of the image tests meant to catch the other bugs, but I'm not clear why my version is worse than the test version...

## Test

![imshow_masked_interpolation-expected](https://user-images.githubusercontent.com/1562854/36575310-98f62542-1807-11e8-943d-d48e9725faee.png)


## Moving the norm step up (pre interp)

![imshow_masked_interpolation](https://user-images.githubusercontent.com/1562854/36575295-875ad648-1807-11e8-8314-d87b70de600f.png)

**Comment by tacaswell** [comment]
The ringing / saturation in the lower left of the test is correct.  If you normalize first you are effectively clipping the impact of the outliers.

**Comment by jklymak** [comment]
Well, we *could* do something like pre-clip the data to some suitably large number around the vmin/vmax, but not 17 orders of magnitude.  That should preserve both behaviours

Or folks could mask their invalid data instead of passing in huge (I assume) invalid values.

**Comment by hagenw** [comment]
I would not assume that huge data points are automatically invalid. In our application we have values between -1 and 1 in most of the image, but it can include regions with values approaching Inf.

I would find it strange to mask some of the data before plotting them. I'm also not aware that you have to do something like this in Matlab or gnuplot.

**Comment by jklymak** [comment]
See #10613 for a fix....

**Comment by tacaswell** [comment]
closed by #10613
